core/market/decentralized/src/db/dao/agreement.rs

```diff
@@ -16,6 +16,14 @@ impl<'a> AsDao<'a> for AgreementDao<'a> {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum StateError {
+    #[error("Can't update Agreement [{0}] state from {1} to {2}.")]
+    InvalidTransition(AgreementId, AgreementState, AgreementState),
```

Comment:
we have two of `AgreementState` here, so I'd go for named fields (struct) instead of a tuple.